File: 프로그래머스/3/12927. 야근 지수/야근 지수.cpp
#include <string>
#include <vector>
#include<algorithm>
#include<unordered_map>
#include<set>
using namespace std;

long long solution(int n, vector<int> works) {
    long long answer = 0;
    unordered_map<int, int> all;
	int high = 0;
	for (auto w : works)
	{
		all[w]++;
		high = max(high, w);
	}

	for (int a = high; a>0;a--)
	{
		if (n == 0) break;

		if (all[a] <= n)
		{
			all[a - 1] += all[a];
			n -= all[a];
			all[a] = 0;

		}
		else if (all[a] > n)
		{
			all[a - 1] += n;
			all[a] -= n;
			n = 0;
		}
	}

	for (auto a : all)
	{
		if (a.second != 0)
		{
			answer += (long long)(a.first * a.first) * a.second;
		}
	}
    return answer;
}

File: 프로그래머스/3/12927. 야근 지수/야근 지수.cs
using System;
using System.Linq;
using System.Collections.Generic;
public class Solution {
    public long solution(int n, int[] works) {
        long answer = 0;
Array.Sort(works);
Array.Reverse(works);
Dictionary<int,int> dict = new Dictionary<int,int>();
for(int i = 0;i<works.Length;i++)
{
    if(!dict.ContainsKey(works[i]))
    {
        dict.Add(works[i], 1);
    }
    else
        dict[works[i]]++;
}

int change = 0;
while (n != 0)
{
    if (dict.Count == 0) break;
    int q = dict.Keys.Max();
    if (dict[q] >= n)
    {
        change = n;
        dict[q] -= n;
        n = 0;
    }
    else
    {
        change = dict[q];
        n -= dict[q];
        dict.Remove(q);
    }

    if (q - 1 != 0)
    {
        if (dict.ContainsKey(q - 1))
        {
            dict[q - 1] += change;
        }
        else dict.Add(q - 1, change);
    }

}

foreach(var item in dict)
{
    if(item.Value != 0)
        answer += ((long)item.Key * item.Key)*item.Value;
}
        return answer;
    }
}